ethcore/src/client/client.rs

```diff
@@ -1853,21 +1853,34 @@ impl BlockChainClient for Client {
 	}
 
 	fn logs(&self, filter: Filter) -> Vec<LocalizedLogEntry> {
-		let (from, to) = match (self.block_number_ref(&filter.from_block), self.block_number_ref(&filter.to_block)) {
-			(Some(from), Some(to)) => (from, to),
-			_ => return Vec::new(),
-		};
+		let blocks = match &filter.from_block {
+			&BlockId::Hash(hash) if filter.from_block == filter.to_block => {
```

Comment:
Seems to be fixing only one very specific case where we query single block by hash?

Reply:
Yes. But that specific case is just what we need to fix the REMOVED-not-displayed bug: https://github.com/paritytech/parity/blob/master/rpc/src/v1/impls/eth_pubsub.rs#L154

I think for other cases where we may want to query branch blocks, we cannot use `from_block` and `to_block` semantic.

Reply:
I'd add a comment explaining that it's pretty much a hack, to support pubsub removed logs notifications then.
I think querying a branch with `from: 5, to: hash(branch of 5)`  should probably work correctly as well.

I still don't understand why it doesn't work. We don't really convert hashes to "canon block numbers", but for every hash we query block header that contains a number - that should work correctly for branch blocks as well, shouldn't it?

Reply:
Spoke with @sorpaas in private chat and he explained me the issues and together we have came up with a different solution to solve this.